Dockerfile.windows

```diff
@@ -249,19 +252,60 @@ RUN `
   Remove-Item C:\binutils.zip; `
   Remove-Item C:\gitsetup.zip; `
   `
-  Write-Host INFO: Creating source directory...; `
-  New-Item -ItemType Directory -Path ${GOPATH}\src\github.com\docker\docker | Out-Null; `
+  # Ensure all directories exist that we will require below....
+  $srcDir = """$Env:GOPATH`\src\github.com\docker\docker\bundles"""; `
+  Write-Host INFO: Ensuring existence of directory $srcDir...; `
+  New-Item -Force -ItemType Directory -Path $srcDir | Out-Null; `
   `
   Write-Host INFO: Configuring git core.autocrlf...; `
-  C:\git\cmd\git config --global core.autocrlf true; `
+  C:\git\cmd\git config --global core.autocrlf true; 
+
+  # Need a different layer  because the current layer  will not
+  # have go in its path. We could always figure that out either
+  # by observing  the current  behavior of  go-s installer (and 
+  # assuming it will not change in the future) or by  observing
+  # change in  PATH value due to the install. Instead we choose
+  # to just build a different layer - arguably an approach that
+  # is more robust.
+RUN `
+  `
+  Function Build-GoTestSum() { `
+    $GotestsumPath=Join-Path -Path $Env:GOPATH -ChildPath "src\gotest.tools\gotestsum"; `
+    Write-Host "INFO: Building gotestsum version $Env:GOTESTSUM_COMMIT in $Env:GOPATH"; `
+    $optsForGet = @('"get"', '"-d"', '"gotest.tools/gotestsum"'); ` 
+    &go $optsForGet; `
+    $savedExitCode = $LASTEXITCODE; `
+    if ($savedExitCode -ne 0) {  `
+      Throw '"Failed getting gotestsum sources..."'  `
+    }; `
+    Write-Host "INFO:     Sources obtained for gotestsum..."; `
+    Push-Location $GotestsumPath; `
+    $optsForCheckout = @('"checkout"', '"-q"', """$Env:GOTESTSUM_COMMIT"""); ` 
+    &git $optsForCheckout; `
+    $savedExitCode = $LASTEXITCODE; `
+    if ($savedExitCode -eq 0) { `
```

Comment:
If you reverse the logic here, we can `Throw` directly here, and probably don't need the `$savedExitCode`. With the changes I suggested, it would look like;

```powershell
  Function Build-GoTestSum() { `
    Write-Host "INFO: Building gotestsum version $Env:GOTESTSUM_COMMIT"; `
    &go get -d gotest.tools/gotestsum; `
    if ($LASTEXITCODE -ne 0) {  `
      Throw '"Failed getting gotestsum sources..."'  `
    }; `
    Write-Host "INFO:     Sources obtained for gotestsum..."; `
    $GotestsumPath=Join-Path -Path $Env:GOPATH -ChildPath "src\gotest.tools\gotestsum"; `
    Push-Location $GotestsumPath; `
    &git checkout -q "$Env:GOTESTSUM_COMMIT"; `
    if ($LASTEXITCODE -ne 0) { `
      Throw '"gotestsum checkout failed..."'; `
    } `
    Write-Host "INFO:     Checkout done for gotestsum..."; `
    &go build --buildmode=exe; `
    if ($LASTEXITCODE -ne 0) {  `
      Throw '"gotestsum build failed..."'; `
    } `
    Pop-Location; `
    Write-Host "INFO:     Build done for gotestsum..."; `
  } `
  `
  Build-GoTestSum
```


Looking further, perhaps if we change `ErrorActionPreference`

```powershell
$ErrorActionPreference = "SilentlyContinue"
```

In which case, we may not even need the manual handling of `LASTEXITCODE`. If I'm right, we could simplify the script quite a lot. Taking if further; given that everything is in a single run, we wouldn't have to worry about `Push-location` / `Pop-location`, and just change to the `$GotestsumPath` directory. At that point, the function is so minimal, I'd even consider just not creating a function for it;

```powershell
  Write-Host "INFO: Building gotestsum version $Env:GOTESTSUM_COMMIT"; `
  &go get -d gotest.tools/gotestsum; `
  Write-Host "INFO:     Sources obtained for gotestsum..."; `
  $GotestsumPath=Join-Path -Path $Env:GOPATH -ChildPath "src\gotest.tools\gotestsum"; `
  Set-Location -Path $GotestsumPath; `
  &git checkout -q "$Env:GOTESTSUM_COMMIT"; `
  Write-Host "INFO:     Checkout done for gotestsum..."; `
  &go build --buildmode=exe; `
  Write-Host "INFO:     Build done for gotestsum...";
```

(even the `INFO` logs may be a bit redundant at that point);

```powershell
  Write-Host "INFO: Building gotestsum version $Env:GOTESTSUM_COMMIT"; `
  &go get -d gotest.tools/gotestsum; `
  Set-Location "$Env:GOPATH\src\gotest.tools\gotestsum"; `
  &git checkout -q "$Env:GOTESTSUM_COMMIT"; `
  &go build --buildmode=exe; `
  Write-Host "INFO: Build done for gotestsum...";
```

Reply:
This is a very good suggestion but I feel it falls outside the scope of current work. I would feel more comfortable making such logical changes when we can thoroughly testing them. So I would suggest that we take this up through other PR-s or cleanup.

Reply:
FWIW; looks like PowerShell 7 will support `&&` to chain commands, which could be an option in future (although likely only for recent base images, so unlikely for RS1); see https://github.com/PowerShell/PowerShell/pull/9849